### ISSUE TITLE
Squiz.WhiteSpace.FunctionSpacing does not detect AfterLast/BeforeFirst for nested functions

### ISSUE THREAD

**Issue by umherirrender** [squizlabs/PHP_CodeSniffer#2406]
The mediawiki-codesniffer standard has a sniff to detect nested functions. It has a test with the following code:
```
$bar = function () {
	{
		/**
		 * Don't touch this
		 */
		function forbiddenToo() { <- ERROR | [x] Expected 1 blank line before function; 0 found (Squiz.WhiteSpace.FunctionSpacing.Before)
			echo 'fail!';
		} <- ERROR | [x] Expected 1 blank line after function; 0 found (Squiz.WhiteSpace.FunctionSpacing.After)
	}
};
class Foo {
	public function returnAnonymousClass() {
		return new class() {
			public function baz() {
				echo 'pass';
			}
		};
	}

	public function bar() {
		function forbidden() {
			echo 'fail!';
		}
	} <- ERROR | [x] Expected 1 blank line after function; 0 found (Squiz.WhiteSpace.FunctionSpacing.After)
}
```

In case of the marked places I would aspect the error code `AfterLast` or `BeforeFirst` instead of only After/Before as is.

I am using spacing = 1 on 3.4.0

**Comment by jrfnl** [comment]
I know one PR with a fix to this sniff is lined up for 3.4.1, but can't test at this moment whether that solves this too (not in the office). If you like and if you can, please have a look at PR #2277 or test with `master` to see if the issue still exists.

**Comment by umherirrender** [comment]
I have tested and there is no difference, so the issue still exists

**Comment by jrfnl** [comment]
@umherirrender Thanks for testing and confirming.

**Comment by gsherwood** [comment]
The easiest way to get this working is to just look for curly braces above or below the function. If one is found (even if it is just a statement group brace with no scope) you assume first/last. This allows all the test cases in this issue to work, and allows all existing unit tests to pass as well.

There is obviously a bit of a change to error codes here, and thus also spacing rules for first/last, but I think this sniff is now more accurate and so that could still be included in 3.5.0.

**Comment by gsherwood** [comment]
Actually I think I'll still limit this to OO scopes, functions, and no-scoped statement groups otherwise it would apply to functions defined inside IF statements as well.

**Comment by gsherwood** [comment]
After my change, the following errors will be reported for the original code:
```
----------------------------------------------------------------------
FOUND 8 ERRORS AFFECTING 8 LINES
----------------------------------------------------------------------
  7 | ERROR | [x] Expected 1 blank line before function; 0 found
    |       |     (Squiz.WhiteSpace.FunctionSpacing.BeforeFirst)
  9 | ERROR | [x] Expected 1 blank line after function; 0 found
    |       |     (Squiz.WhiteSpace.FunctionSpacing.AfterLast)
 13 | ERROR | [x] Expected 1 blank line before function; 0 found
    |       |     (Squiz.WhiteSpace.FunctionSpacing.BeforeFirst)
 15 | ERROR | [x] Expected 1 blank line before function; 0 found
    |       |     (Squiz.WhiteSpace.FunctionSpacing.BeforeFirst)
 17 | ERROR | [x] Expected 1 blank line after function; 0 found
    |       |     (Squiz.WhiteSpace.FunctionSpacing.AfterLast)
 22 | ERROR | [x] Expected 1 blank line before function; 0 found
    |       |     (Squiz.WhiteSpace.FunctionSpacing.BeforeFirst)
 24 | ERROR | [x] Expected 1 blank line after function; 0 found
    |       |     (Squiz.WhiteSpace.FunctionSpacing.AfterLast)
 25 | ERROR | [x] Expected 1 blank line after function; 0 found
    |       |     (Squiz.WhiteSpace.FunctionSpacing.AfterLast)
----------------------------------------------------------------------
PHPCBF CAN FIX THE 8 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```

**Comment by gsherwood** [comment]
This change has now been committed. Thanks for reporting this issue.
